[buck2][caffe2] stable paths in generated file comments

johnhaitas opened the pull request:
Summary:
We don't want absolute paths in generated file output because that will change from machine to machine and not be properly cacheable.

### Before
> // generated from /data/sandcastle/boxes/eden-trunk-hg-fbobjc-fbsource/buck-out/gen/03598924/xplat/caffe2/tools/setup_helpers/__generate_code_bin#link-tree/tools/autograd/templates/VariableType.h

{F739745848} {F739746074} {F739746230}

### After
> // generated from tools/autograd/templates/VariableType.h

{F739753190} {F739753374} {F739753519}

Test Plan:
### buck1
```
buck1 build -c cxx.available_platforms=CXX,APPLE fbsource//xplat/caffe2:gen_aten_libtorch
```

### buck2
```
buck2 build -c cxx.available_platforms=CXX,APPLE fbsource//xplat/caffe2:gen_aten_libtorch
```

Differential Revision: D36960456

dagitses commented on tools/autograd/gen_autograd_functions.py:
```diff
@@ -346,6 +346,10 @@
 # TODO: This is probably not exhaustive, but it's a start
 UNTRACEABLE_FUNCTIONS = VIEW_FUNCTIONS
 
+_HASH_LINK_TREE = "#link-tree/"
```
instead of having something Buck specific here, how about just getting the relative path from above the tools directory?

We can identify the tools directory with `pathlib.Path(__file__).parent.parent`

dagitses reviewed (changes requested):
This is a great idea. But make sure it works in opt mode too since code gen will run from a par file, I believe.

johnhaitas committed:
[buck2][caffe2] stable paths in generated file comments (#79133)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79133

We don't want absolute paths in generated file output because that will change from machine to machine and not be properly cacheable.

### Before
> // generated from /data/sandcastle/boxes/eden-trunk-hg-fbobjc-fbsource/buck-out/gen/03598924/xplat/caffe2/tools/setup_helpers/__generate_code_bin#link-tree/tools/autograd/templates/VariableType.h

{F739745848} {F739746074} {F739746230}

### After
> // generated from tools/autograd/templates/VariableType.h

{F739753190} {F739753374} {F739753519}

Test Plan:
### buck1
```
buck1 build -c cxx.available_platforms=CXX,APPLE fbsource//xplat/caffe2:gen_aten_libtorch
```

### buck2
```
buck2 build -c cxx.available_platforms=CXX,APPLE fbsource//xplat/caffe2:gen_aten_libtorch
```

Reviewed By: ezyang

Differential Revision: D36960456

fbshipit-source-id: d2ed46a0311eda84557dead26cd8b316fc43a177
